lib/usdUfe/ufe/UsdClipboardCommands.cpp

```diff
@@ -0,0 +1,496 @@
+
+        // Set the material parent name in the ClipboardMetadata.
+        setClipboardMetadata(usdItem, std::string(kMaterialName), materialName);
+
+        // Set the stage path in the ClipboardMetadata.
+        setClipboardMetadata(usdItem, std::string(kStagePath), stagePath);
+    }
+}
+
+void clearClipboardMetadata(Ufe::SceneItemList& targetItems)
+{
+    std::shared_ptr<Ufe::CompositeUndoableCommand> compositeClearMetadataCmd
+        = std::make_shared<Ufe::CompositeUndoableCommand>();
+
+    // Remove ClipboardMetadata.
+    for (auto& targetItem : targetItems) {
+        auto usdItem = std::dynamic_pointer_cast<UsdSceneItem>(targetItem);
+
+        if (PXR_NS::UsdShadeMaterial(usdItem->prim()))
+            continue;
+
+        compositeClearMetadataCmd->append(
+            usdItem->clearGroupMetadataCmd(std::string(kClipboardMetadata)));
+    }
+
+    compositeClearMetadataCmd->execute();
+}
+
+Ufe::SceneItem::Ptr renameItemUsingMetadata(Ufe::SceneItem::Ptr item)
+{
+    if (item) {
+        auto usdItem = std::dynamic_pointer_cast<UsdSceneItem>(item);
+        if (usdItem) {
+            const auto newName
+                = usdItem->getGroupMetadata(std::string(kClipboardMetadata), std::string(kNodeName))
+                      .get<std::string>();
+
+            // Nothing to rename, return the item.
+            if (item->nodeName() == newName || newName.empty()) {
+                return item;
+            }
+
+            const Ufe::SceneItemOps::Ptr sceneItemOps = Ufe::SceneItemOps::sceneItemOps(item);
+            if (sceneItemOps) {
+                Ufe::PathComponent namePathComponent(newName);
+                const auto renameCmd = sceneItemOps->renameItemCmdNoExecute(namePathComponent);
+                if (renameCmd) {
+                    renameCmd->execute();
+                    return renameCmd->sceneItem();
+                }
+            }
+        }
+    }
+
+    return {};
+}
+
+Ufe::SceneItemList pasteItemsToNewMaterial(const UsdSceneItem::Ptr& dstItem, Ufe::Selection& items)
+{
+    auto compositeCmd = Ufe::CompositeUndoableCommand::create({});
+
+    // Preserve the scene item order from the selection.
+    std::map<std::string, std::map<std::string, Ufe::Selection>> stageMaterialNamesMap;
+    Ufe::SceneItemList                                           createdMaterials;
+
+    // Group all the items with the same material name and from the same stage.
+    for (const auto& item : items) {
+        const auto materialName
+            = item->getGroupMetadata(std::string(kClipboardMetadata), std::string(kMaterialName))
+                  .get<std::string>();
+        const auto originStage
+            = item->getGroupMetadata(std::string(kClipboardMetadata), std::string(kStagePath))
+                  .get<std::string>();
+
+        stageMaterialNamesMap[originStage][materialName].append(item);
+    }
+
+    // Create the new Materials taking into consideration also the stage.
```

Comment:
As I understand it, the goal of this function is to create new materials and put a bunch of items that need to be pasted under them? One materials is created for all items that were under the same material originally? Maybe that should be a command at the start or above the function.